src/OrchardCore.Modules/OrchardCore.Media/Processing/MediaTokenService.cs

```diff
@@ -98,30 +131,97 @@ public bool TryValidateToken(IDictionary<string, string> commands, string token)
             return false;
         }
 
-        private static string CreateQueryStringTokenKey(IEnumerable<string> values)
+        /// <summary>
+        /// Specialized version with fast enumeration and no StringValues string allocation.
+        /// </summary>
+        private static string CreateQueryStringTokenKey(Dictionary<string, StringValues> values)
+        {
+            using var builder = ZString.CreateStringBuilder();
+            builder.Append(TokenCacheKeyPrefix);
+            foreach (var pair in values)
+            {
+                builder.Append(pair.Value.ToString());
+            }
+            return builder.ToString();
+        }
+
+        private static string CreateQueryStringTokenKey(IDictionary<string, string> values)
         {
             using var builder = ZString.CreateStringBuilder();
             builder.Append(TokenCacheKeyPrefix);
-            foreach (var item in values)
+            foreach (var pair in values)
             {
-                builder.Append(item);
+                builder.Append(pair.Value);
             }
             return builder.ToString();
         }
 
         private string GetHash(string queryStringTokenKey)
-            => _memoryCache.GetOrCreate(queryStringTokenKey, entry =>
-               {
-                   entry.SlidingExpiration = TimeSpan.FromHours(5);
+        {
+            if (!_memoryCache.TryGetValue(queryStringTokenKey, out var result))
```

Comment:
old code had capturing closure as extension method took a Func which needed outside context, now using methods actually present in `IMemoryCache` to prevent that